src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_grand_champions.cpp

```diff
@@ -182,75 +319,215 @@ class generic_vehicleAI_toc5 : public CreatureScript
-                    AddWaypoint(0, 747.35f, 634.07f, 411.57f);
-                    AddWaypoint(1, 768.72f, 581.01f, 411.92f);
-                    AddWaypoint(2, 763.55f, 590.52f, 411.71f);
+                    if (uiData == HORDE)
+                    {
+                        AddWaypoint(0, 747.35f, 634.07f, 411.57f);
+                        AddWaypoint(1, 792.36f, 614.47f, 412.39f);
+                        AddWaypoint(2, GrandChampFinalLoc[4].GetPositionX(), GrandChampFinalLoc[4].GetPositionY(), GrandChampFinalLoc[4].GetPositionZ());
+                    }
+                    else if (uiData == ALLIANCE)
+                    {
+                        AddWaypoint(0, 747.35f, 634.07f, 411.57f);
+                        AddWaypoint(1, 708.31f, 635.07f, 411.93f);
+                        AddWaypoint(2, GrandChampFinalLoc[1].GetPositionX(), GrandChampFinalLoc[1].GetPositionY(), GrandChampFinalLoc[1].GetPositionZ());
+                    }
                     uiWaypointPath = 2;
                     break;
                 case 3:
-                    AddWaypoint(0, 747.35f, 634.07f, 411.57f);
-                    AddWaypoint(1, 784.02f, 645.33f, 412.39f);
-                    AddWaypoint(2, 775.67f, 641.91f, 411.91f);
+                    if (uiData == HORDE)
+                    {
+                        AddWaypoint(0, 747.35f, 634.07f, 411.57f);
+                        AddWaypoint(1, 793.00f, 624.40f, 412.39f);
+                        AddWaypoint(2, GrandChampFinalLoc[5].GetPositionX(), GrandChampFinalLoc[5].GetPositionY(), GrandChampFinalLoc[5].GetPositionZ());
+                    }
+                    else if (uiData == ALLIANCE)
+                    {
+                        AddWaypoint(0, 747.35f, 634.07f, 411.57f);
+                        AddWaypoint(1, 716.67f, 651.1f, 411.93f);
+                        AddWaypoint(2, GrandChampFinalLoc[2].GetPositionX(), GrandChampFinalLoc[2].GetPositionY(), GrandChampFinalLoc[2].GetPositionZ());
+                    }
                     uiWaypointPath = 3;
                     break;
+                case 4:
+                    AddWaypoint(0, LesserChampLoc[uiData].GetPositionX(), LesserChampLoc[uiData].GetPositionY(), LesserChampLoc[uiData].GetPositionZ());
+                    uiWaypointPath = 4;
+                    Start(false); // walking
+                    break;
             }
 
             if (uiType <= 3)
                 Start(false, true);
         }
 
-        void WaypointReached(uint32 waypointId) override
+        void WaypointReached(uint32 waypointId) override 
         {
-            switch (waypointId)
+            uint32 TeamInInstance = instance->GetData(DATA_PLAYERS_TEAM);
+            // Grand Champions reached their waiting positions in the jousting event
+            // and Lesser Champions begin to correct their positions
+            if ((uiWaypointPath == 1 && waypointId == 3) || ((uiWaypointPath == 2 || uiWaypointPath == 3) && waypointId == 2))
```

Comment:
please use enums or defines or anything else instead of magic number like 1,2,3 that don't really tell much about what these checks do